api/METRICS.md

```diff
@@ -86,21 +86,186 @@ group_max_from_summary(summary)
+
+**Status quo:**
+
+* In both cases, `objective` is automatically inferred from the provided `constraints`.
+
+* `ExponentiatedGradient(estimator, constraints, eps=`&epsilon;`)` considers constraints that are specified jointly by the provided `Moment` object and the numeric value &epsilon; as follows:
+
+  * `DemographicParity()`: for all _a_, <br>
+    |_selection_rate_(_a_) - _selection_rate_(\*)| &le; &epsilon;.
+
+  * `DemographicParity(ratio=`_r_`)`: for all _a_, <br>
+    _r_ &sdot; _selection_rate_(_a_) - _selection_rate_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _selection_rate_(\*) - _selection_rate_(_a_) &le; &epsilon;.
+
+  * `TruePositiveRateDifference()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;.
+
+  * `TruePositiveRateDifference(ratio=`_r_`)`: analogous
+
+  * `EqualizedOdds()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;, <br>
+    |_false_positive_rate_(_a_) - _false_positive_rate_(\*)| &le; &epsilon;.
+
+  * `EqualizedOdds(ratio=`_r_`)`: analogous
+
+  * `ErrorRateRatio()`: for all _a_, <br>
+    |_error_rate_(_a_) - _error_rate_(\*)| &le; &epsilon;.
+
+  * `ErrorRateDifference(ratio=`_r_`)`: analogous
+
+  * all of the above constraints are descendants of `ConditionalSelectionRate`
+
+  * the `objective` for all of the above constraints is `ErrorRate()`
+
+* `GridSearch(estimator, constraints)` considers constraints represented by the provided `Moment` object. The behavior of the `GridSearch` algorithm does not depend on the value of the right-hand side of constraints, so it is not provided to the constructor. In addition to the `Moment` objects above, `GridSearch` also supports the following:
+
+  * `GroupLossMoment(<loss>)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    * where `<loss>` is a _loss evaluation_ object; it supports an interface that takes `y_true` and `y_pred` as the input and returns the vector of losses evaluated on individual examples
+
+    * the `objective` for `GroupLossMoment(<loss>)` is `AverageLossMoment(<loss>)`
+
+  * both `GroupLossMoment` and `AverageLossMoment` are descendants of `ConditionalLossMoment`
+
+**Proposal.** The proposal introduces some breaking changes:
+
+* `constraints`:
+
+  * `<Metric>Parity(difference_bound=`&epsilon;`)`: for all _a_, <br>
+    |_metric_(_a_) - _metric_(\*)| &le; &epsilon;.
+
+  * `<Metric>Parity(ratio_bound=`_r_`, ratio_bound_slack=`&epsilon;`)`: for all _a_, <br>
+    _r_ &sdot; _metric_(_a_) - _metric_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _metric_(\*) - _metric_(_a_) &le; &epsilon;.
+
+  * `DemographicParity` and `EqualizedOdds` have the same calling convention as `<Metric>Parity`
+
+  * `BoundedGroupLoss(<loss>, bound=`&zeta;`)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    _Alternative proposals_:
+    * `LossParity(<loss>, group_max_bound=`&zeta;`)` <br>

```

Comment:
I like the generic `LossParity(<loss>, …)` better, but will the internal logic be identical with only the loss definition being the difference? If not, then that's a good case for having the individual `<Loss>Parity(…)` classes

Reply:
There would be no additional logic within `<Loss>Parity` if we were only going to support `group_max_bound`.

The main reason why I prefer `BoundedGroupLoss` over `LossParity`, is because in my mind "parity" indicates equality, whereas the BGL criterion is one sided.

Reply:
How about `LossConstraint`? That's more generic.

Reply:
Is there a reason you don't like `BoundedGroupLoss`? It's how we call it in the paper and it is reasonably descriptive. I don't necessarily love it, but alternatives have their issues the. The issue with `LossConstraint` is that it does not capture the fact that group-level losses are meant there.

Reply:
I don't mind it. I just thought you were trying to create something more generic. Either way is fine :-)

Reply:
Ok then. Then let's stick to `BoundedGroupLoss` since that's the terminology that we used in the paper--and it's also reasonably descriptive.